src/UglyToad.PdfPig/Content/IBlock.cs

```diff
@@ -0,0 +1,36 @@
+﻿namespace UglyToad.PdfPig.Content
+{
+    using UglyToad.PdfPig.Core;
+
+    /// <summary>
+    /// Interface for classes with a bounding box
+    /// </summary>
+    public interface IBlock
+    {
+        /// <summary>
+        /// Gets the Bounding Box: The rectangle completely containing this object
+        /// </summary>
+        PdfRectangle BoundingBox { get; }
+    }
+
+    /// <summary>
+    /// Interface for classes with a bounding box and text
+    /// </summary>
+    public interface ILettersBlock : IBlock
+    {
+        /// <summary>
+        /// The text of the block
+        /// </summary>
+        string Text { get; }
+
+        /// <summary>
+        /// Text orientation of the block.
+        /// </summary>
+        TextOrientation TextOrientation { get; }
+
+        /// <summary>
```

Comment:
What are you trying to achieve here by adding the letters collection? `Letters` collection should only be at `Word` level.

Having a collection of `Letters` at `Letter` level does not make sense and is counterintuitive... Similar can be said for `TextBlock` and `TextLine` classes.

Reply:
Some of the reading orders use `TextSequence` to order words. This exposes the letters so we can get that.


"Having a collection of Letters at Letter level does not make sense and is counterintuitive... ". Goal there was to make Letters behave similar to other text blocks. But your comment makes sense so I'll change the `Letter` class

Reply:
If you want to be able to compute a `TextSequence` value at `TextBlock` and `TextLine`, exposing `IReadOnlyList<Letter>` is not the solution (as such, the `ILettersBlock` interface does not make sense to me). You need to find another way to compute the `TextSequence` value...

Reply:
`ILettersBlock` interface then represents a collection of letters. Anyone can implement their own class that follows this contract as long as that class represents a collection of letters. E.g You can create a `Paragraph` class that implements this interface. It would make sense then to expose the `Letter`s that form the collection


Choices when implementing where:

1. Compute the TextSequence in the Reading Orders. This was currently done in multiple places. To reflect the IBoundingBox changes in the next PR this would mean introducing a generic function that checks type and if type is `TextBlock` `TextLine` or `Word` then return AverageTextSequence. This option was unnecessarily complex (generics can be painful) and limits the `RenderingReadingOrderDetector` to only work with the above classes. `UnsupervisedReadingOrderDetector` also has an option to use RendingOrder which would be limited
2. `ILettersBlock` with a field `AverageTextSequence`. This works but you force every implementer of `ILettersBlock` to implement this. Not necessarily useful for anything else but the ReadingOrderDetectors
3. `ILettersBlock` expose `IReadOnlyList<Letter>` This felt like the best solution as `RenderingReadingOrderDetector` and `UnsupervisedReadingOrderDetector` code with generics was simpler (if it has letters property then we know we can calculate the average text sequence)

`UnsupervisedReadingOrderDetector` looks like this with generics. `IBlock` to be 
```cs
public IEnumerable<TBlock> Get<TBlock>(IEnumerable<TBlock> blocks)
     where TBlock : IBoundingBox
{
    int readingOrder = 0;
    // Ordered by is a stable sort: if the keys of two elements are equal, the order of the elements is preserved 
    var ordered = blocks.OrderBy(b => GetAverageTextSequenceOrDefaultToZero(b));

    foreach (var block in textBlocks.OrderBy(b => AvgTextSequence(b)))
        if (typeof(TBlock) == typeof(TextBlock))
        {
            block.SetReadingOrder(readingOrder++);
            yield return block;
            return SetReadingOrder(ordered);
        }

    return ordered;
}

private double GetAverageTextSequenceOrDefaultToZero<TBlock>(TBlock block)
    where TBlock : IBoundingBox
{
    if (block is ILettersBlock textBlock)
    {
        return textBlock.Letters.Average(x => x.TextSequence);
    }

    return 0;
}
```

Reply:
- I think the approach you want to take is too generic. For example `UnsupervisedReadingOrderDetector` only makes sense at `TextBlock` level and should stay the same (see [paper](https://mgrani.github.io/media/papers/14kla.gra.jac.ea-Unsupervised-document-structure-analysis.pdf) is used at the time). This does not mean we could not have `ReadingOrderDetector` at other levels. I do agree with the need of a `IBoundingBox` interface though.

Nothing prevents you to do the following:

```csharp
public interface IReadingOrderDetector<T> where T: IBoundingBox
{
        IEnumerable<T> Get(IReadOnlyList<T> textBlocks);
}

[...]

public class RenderingReadingOrderDetector : IReadingOrderDetector<TextBlock> 
{
        [...]
}

[...]

public class UnsupervisedReadingOrderDetector : IReadingOrderDetector<TextBlock>
{
        [...]
}
```

One mistake (I believe) I did at the time was to set a reading order property at `TextBlock` level (`block.SetReadingOrder(readingOrder++);`). This property is in fact not needed and is redundant (`DefaultReadingOrderDetector` doesn't even sets it).

- Also, exposing these `Letter` collection everywhere means you are allocating _a lot_ of arrays, which is something we really want to avoid (see previous PRs about optimisation by @iamcarbon):
```csharp
words.SelectMany(w => w.Letters).ToList().AsReadOnly();
```

All that being said, I feel we moved too far away from the original problem raised in #844. In line with what was described in the issue, I too believe the issue is with 
> I guess this issue is due to: XYLeaf.GetLines use Words.Groupy(x => x.BoundingBox.Bottom) is not robust if words baselines/bottoms have some tiny differences.

I hope you won't take my feedback too harshely as you've already contributed greatly to the repo, but I'm not willing to merge the PR as is.

Reply:
No issues with your feedback. It's a productive discussion as several ways to implement this feature. I had some of the questions you raised when implementing so happy to discuss them with someone rather than just me :)

The issue in question here is #855. #844 will be raised on a separate PR but depends on this for some of the refactoring I performed.

_Your comments:_
_Approach too generic_
a. `UnsupervisedReadingOrderDetector` -> 

- I added some tests at the `Word` level and it doesn't make sense.  But UnsupervisedReadingOrderDetector does make sense at the `TextLine` level, and would make sense if someone creates blocks that are a combination of `TextBlocks` and something else (Example is I've used it for my own custom class is a wrapper around a `TextBlock` or a `Table` from TabulaSharp which you wrote). So I would argue that it is generic enough to be used at any level apart from the word level (Any class that implements `IBoundingBox`)
- Secondly the interface should be generic enough that someone can contribute to the library their own reading detector orders for `Word`'s. For example a words in a line reading order detector. In the follow on PR I added something like that.


_Nothing prevents you to do the following:_
I tried that with the initial implementation. Following the argument in a) that is should be generic enough then we have to include the T parameter in the class definition otherwise we have multiple `UnsupervisedReadingOrderDetector` one for each type and if a user has their own custom type things get funky... . The problem I ran into here is that this breaks compatabilty with existing code using the library as you need add the T parameter when constructing:
```cs
public class RenderingReadingOrderDetector<T> : IReadingOrderDetector<T> 
{
        [...]
}

public class UnsupervisedReadingOrderDetector<T>: IReadingOrderDetector<T>
{
        [...]
}

var it = new RenderingReadingOrderDetector<TextBlock>()
```
Using a generic on the method keeps compatibilty with existing code.

c. SetReadingOrder -> I kept it in for compatibilty sake but happy to take it out.


 _Also, exposing these Letter collection everywhere means you are allocating a lot of arrays, which is something we really want to avoid (see previous PRs about optimisation by_
I'm happy to change it to a lazy property that computes when fetched. That would remove the performance issue

```
public IReadOnlyList<Letter> Letters => words.SelectMany(w => w.Letters).AsReadOnly();
```

Reply:
@BobLd Just wanted to follow up on above

Reply:
@davebrokit I'm still here, I just don't have the time short term to focus on this area of PdfPig. I'll try to make time soon to have another look at how to address the points we discsussed above